packages/flutter_tools/lib/src/dart/pub.dart

```diff
@@ -477,17 +522,58 @@ class _DefaultPub implements Pub {
     return values.join(':');
   }
 
-  String? _getRootPubCacheIfAvailable() {
+  /// There is 3 ways to get the pub cache location
+  ///
+  /// 1) Provide the _kPubCacheEnvironmentKey.
+  /// 2) localPath (flutterRoot) but not globalPath ($HOME) .pubCache in which case we use the available (and vice versa).
+  /// 3) If both local and global are available joinCaches is called and local deleted
+  String? _getLocalPubCacheIfAvailable() {
     if (_platform.environment.containsKey(_kPubCacheEnvironmentKey)) {
       return _platform.environment[_kPubCacheEnvironmentKey];
     }
 
-    final String cachePath = _fileSystem.path.join(Cache.flutterRoot!, '.pub-cache');
-    if (_fileSystem.directory(cachePath).existsSync()) {
-      _logger.printTrace('Using $cachePath for the pub cache.');
-      return cachePath;
+    final String localCachePath = _fileSystem.path.join(Cache.flutterRoot!, '.pub-cache');
+    final Directory globalDirectory;
+    if (_platform.isWindows) {
+      // %LOCALAPPDATA% is preferred as the cache location over %APPDATA%, because the latter is synchronised between
+      // devices when the user roams between them, whereas the former is not.
+      // The default cache dir used to be in %APPDATA%, so to avoid breaking old installs,
+      // we use the old dir in %APPDATA% if it exists. Else, we use the new default location
+      // in %LOCALAPPDATA%.
+      final String appData = _platform.environment['APPDATA'] ?? '';
+      final Directory appDataDir = _fileSystem.directory(_fileSystem.path.join(appData, 'Pub', 'Cache'));
+
+      if (appDataDir.existsSync()) {
+        globalDirectory = appDataDir;
+      } else {
+        final String localAppData = _platform.environment['LOCALAPPDATA'] ?? '';
+        globalDirectory = _fileSystem.directory(_fileSystem.path.join(localAppData, 'Pub', 'Cache'));
+      }
+    }
+    else {
+      final String homeDirectory = _platform.environment['HOME'] ?? '';
```

Comment:
no, it is valid to not have a home directory. Maybe `globalDirectory` should be nullable, and in this case it would be null. And then later you treat a `null` `globalDirectory` the same as if it didn't exist.